app/code/Magento/NewsletterApi/Exception/AlreadySubscribedException.php

```diff
@@ -0,0 +1,14 @@
+<?php
+/**
```

Comment:
@adragus-inviqa 
> Well, you're suggesting us to use M2's exceptions, when you guys have created your own? Not fair.

Actually not, we are trying not to introduce custom exception types in Magento modules as well. And rely on \Magento\Framework\Exception\*
(successors of LocalizedException).

One of the reason - is that it doesn't make sense to have an Exception Interface, but it make sense to have just Exception implementation. But logic behind our Service Contracts is to expose API interfaces only. Without any implementation.

Also, practice of creation a lot of custom exceptions will lead to building a business logic on exception throwing/catching for business invariants implementation instead of usage of conditional logic with business statuses etc.
Business logic (both application and domain) MUST NOT be managed with exceptions. Conditional statements should be used instead.

@davidverholen 
> I think the Newsletter Repository should throw the given internal Exceptions, as there are already plenty of good / speaking exceptions for the persistence layer itself.

That's not correct, on each layer should be defined own exceptions hierarchy.
Exception hierarchy for Service Contract level is successors of LocalizedException. It is allowed to throw exceptions only which are defined on the same layer.
If an exception is caught on the layer which is different from the one where it has been thrown, and it should be re-thrown, then a new exception instance appropriate to the current layer should be created (in this case, the original exception must be passed to a new instance with "previous" argument).

For this specific case I don't recommend to use LocalizedException , because it's too generic and  COULD only be thrown in presentation layer (Controllers, Blocks).

For this case, I propose to use `Magento\Framework\Exception\StateException`.  Potentially we could consider to add `\Magento\Framework\Exception\AlreadyExistsException`, but as for me, it's a subtype of StateException.